### PR TITLE
PDCL-1245 - Support redirect Personalization offer types

### PR DESCRIPTION
**Pull request by ninaceban**
## Description

Redirect offers are usually the experiences that customer set to redirect from the experience's entry criteria URL to another URL and if `renderDecisions` is true we'd like to execute it as quick as possible. If there will be more than one `redirect decision` then we will execute the first one. After the redirect decision was executed we will send a notification call. 

- If `renderDecisions` is true we will execute the first redirect decision. 
- If `renderDecisions` is false we will return these decisions to the customer at the `sendEvent`.

If the Redirect Offer is A4T enabled, the redirect URL will contain analytics `tnt_a` encoded payload in the `alloy_redirect` query string param. Analytics data exchange will happen on the Konductor side.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (non-breaking change which does not add functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html) or I'm an Adobe employee.
- [ ] I have made any necessary test changes and all tests pass.
- [ ] I have run the Sandbox successfully.

### CODE REVIEW

**Comment by jfkhoury** [comment]
@ninaceban Can we please add this feature to the sandbox? thanks!

**Review by jonsnyder — APPROVED** [review]

File: src/components/Personalization/createRedirectHandler.js
Comment (Aaronius) [inline src/components/Personalization/createRedirectHandler.js:27]: I'm noticing this in the [docs for `location.replace`](https://developer.mozilla.org/en-US/docs/Web/API/Location/replace):

> If the assignment can't happen because of a security violation, a DOMException of the SECURITY_ERROR type is thrown. This happens if the origin of the script calling the method is different from the origin of the page originally described by the Location object, mostly when the script is hosted on a different domain.

I haven't verified, but that seems to suggest that if Alloy is loaded from a different origin than the page loading Alloy, that an error will be thrown when Alloy calls `location.replace`. At least, that's my interpretation. Have you tested this scenario?

**Review by Aaronius — APPROVED** [review]
I have a concern about an error being thrown when `location.replace` is called, which I added a comment about. Otherwise, the rest is just small tweaks. Nice work @ninaceban.

File: src/components/Personalization/createOnResponseHandler.js
Comment (ninaceban) [reply src/components/Personalization/createOnResponseHandler.js:46]: yes, if we have at least one redirect offer and we want to render it (renderDecisions=true) then it doesn't make any sense to store the view related offers. If we don't want to render them then we will store SPA offers and will resolve the sendEvent promise with `__view__` and `REDIRECT_ITEM` decisions.

File: test/functional/specs/Personalization/C205528.js
Comment (ninaceban) [reply test/functional/specs/Personalization/C205528.js:52]: yes sure! thank you

**Comment by ninaceban** [comment]
> @ninaceban Can we please add this feature to the sandbox? thanks!
sure!

File: src/components/Personalization/createOnResponseHandler.js
Comment (ninaceban) [reply src/components/Personalization/createOnResponseHandler.js:54]: ok, I'll rename it

File: src/components/Personalization/createCollect.js
Comment (ninaceban) [reply src/components/Personalization/createCollect.js:14]: ok, I'll refactor

File: src/components/Personalization/createRedirectHandler.js
Comment (ninaceban) [reply src/components/Personalization/createRedirectHandler.js:21]: thank you, I will refactor it

File: src/components/Personalization/createOnResponseHandler.js
Comment (ninaceban) [reply src/components/Personalization/createOnResponseHandler.js:13]: These are exactly the schema names. The only difference is that the ones defined by DOM_ACTION, are actions that we can execute against DOM, while the 3 others are items that the customer knows how to handle, and based on a schema they have a slightly different structure or goals.

File: src/components/Personalization/createRedirectHandler.js
Comment (ninaceban) [reply src/components/Personalization/createRedirectHandler.js:23]: Thank you, I really appreciate it! I will do a refactor in this area, maybe as part of another PR.

File: src/components/Personalization/createRedirectHandler.js
Comment (ninaceban) [reply src/components/Personalization/createRedirectHandler.js:27]: a try/catch wouldn't cause any harm, but currently, we don't try-catch in at.js. I will add a try/catch, it would catch exceptions like malformed URL as well.

Comment (Aaronius) [reply src/components/Personalization/createRedirectHandler.js]: I'm more concerned that the navigation won't actually occur. You've verified that the navigation actually occurs in this scenario though? I'm probably just misunderstanding MDN's documentation, but now I want to know what they're documentation is trying to say. If no error is being thrown in this scenario and it's actually navigating the user, then there's no need to add a catch block unless we know what/why we're catching.

File: sandbox/src/RedirectOffers.js
Comment (Aaronius) [inline sandbox/src/RedirectOffers.js:11]: This is going to send an event every time this component renders, which probably isn't really what you're after (the component can re-render many times while the user is on the view). If you want to send an event every time the user navigates to this view, you'll probably need to use the `useEffect` React hook. Something like:

```
useEffect(() => {
    window
      .alloy("sendEvent", {
        renderDecisions: true
      })
      .then(({ decisions = [] }) => {
        console.log("personalized decisions", decisions);
      });
  }, []);
  ```

File: test/unit/specs/components/Personalization/createRedirectHandler.spec.js
Comment (Aaronius) [inline test/unit/specs/components/Personalization/createRedirectHandler.spec.js:55]: This description should probably mention something about an error being thrown when attempting to redirect.

File: src/components/Personalization/createRedirectHandler.js
Comment (Aaronius) [reply src/components/Personalization/createRedirectHandler.js]: Ignore my last sentence. It might still valuable to catch. We can leave it. Is it a problem that if we try to redirect and it failes, that we show container but won't have stored any of the views in the cache?

**Review by Aaronius — APPROVED** [review]
I suggested some tweaks. You probably want to fix the redirect offer component not to send an event on every render.

**Review by jfkhoury — APPROVED** [review]
Awesome work @ninaceban !! ✌️  I will follow up with the testing group.

**Comment by Aaronius** [comment]
Let's wait until Eric and others finish testing this feature before merging this PR.

**Comment by jfkhoury** [comment]
@ninaceban Latest Update: We tested this feature and it seems like there's a disconnect between Target & Analytics. We will pull this feature from the release and start design meetings to finish it.

PLEASE DON'T MERGE.

File: sandbox/src/App.js
Comment (Aaronius) [inline sandbox/src/App.js:66]: You'll probably need to update the menu found within legacy.html as well. Keeping the menu in sync in these two files is a chore, but hopefully will be improved in https://jira.corp.adobe.com/browse/PDCL-4822.

File: test/unit/specs/components/Personalization/createRedirectHandler.spec.js
Comment (Aaronius) [inline test/unit/specs/components/Personalization/createRedirectHandler.spec.js:65]: It might be good to assert that `collect` was called with `{ meta, documentMayUnload }`.

**Review by Aaronius — APPROVED** [review]
Nice work @ninaceban!

As a side note, there might be a valid reason for having force-pushed your commits, but I would recommend trying to avoid force pushing unless you're reasonably confident that nobody has looked at your PR yet. Force pushing makes it harder for reviewers to know what has changed since the last time they reviewed, among other things. If you're not sure how to accomplish whatever you were trying to do without force-pushing, let me know and we can talk through it.

Thanks Nina!

File: src/components/Personalization/decisionsExtractor.js
Comment (jonsnyder) [inline src/components/Personalization/decisionsExtractor.js:33]: I'd love to do a refactor here. Not important for this PR, but I think it would simplify this code and make it easier to understand. If you think this will work, I'll create a JIRA ticket.

```
const createDecision = (decision, items) => {
  return { id: decision.id, scope: decision.scope, items };
 };

const findMatchingIndex = (decision, item, predicates) => {
  return predicates.reduce((matchingIndex, predicate, currentIndex) => {
    // Check if we already matched one of the predicates
    if (matchingIndex !== predicates.length) {
      return matchingIndex;
    }
    // Check if this predicate matches
    if (predicate({ decision, item })) {
      return currentIndex;
    }
    // This predicate doesn't match so return index for unmatched items
    return predicates.length;
  }, predicates.length);
}

// returns an array of decisions one greater than the number of predicates.
// the predicates are functions that return true or false, and are given {decision, item}
// index 0 holds the decisions that match the first predicate
// index 1 holds the decisions that match the second predicate
// and so forth
// the last index holds the decisions that didn't match any predicate
const extractDecisions = (decisions, ...predicates) => {
  const matchedDecisions = new Array(predicates.length + 1).map(() => []);
  decisions.forEach(decision => {
    const matchedItems = new Array(predicates.length + 1).map(() => []);
    // note: if a decision doesn't have items, it will be dropped.
    const { items = [] } = decision;
    items.forEach(item => {
      matchedItems[findMatchingIndex(decision, item, predicates)].push(item);
    });
    matchedItems.forEach((matchedItemArray, i) => {
      if (isNonEmptyArray(matchedItemArray)) {
        matchedDecisions[i].push(createDecision(decision, matchedItemArray));
      }
    });
  });
  return matchedDecisions;
};

// inside createOnResponseHandler
const [redirectItems, pageWideDomActions, nonPageWideDomActions, otherDecisions] =
  extractDecisions(
    unprocessedDecisions,
    ({ item }) => item.schema === REDIRECT_ITEM,
    ({ item, decision }) => item.schema === DOM_ACTION && decision.scope === PAGE_WIDE_SCOPE,
    ({ item, decision }) => item.schema === DOM_ACTION && decision.scope !== PAGE_WIDE_SCOPE
  );

```

**Review by jonsnyder — CHANGES_REQUESTED** [review]
We need to use the proposition XDM mixin to record redirects.

File: test/unit/specs/components/Personalization/createRedirectHandler.spec.js
Comment (ninaceban) [reply test/unit/specs/components/Personalization/createRedirectHandler.spec.js:65]: I did that in the test above, I just wanted to assert here that when the redirect itself will throw an error we would show containers and log a warning.

File: sandbox/src/App.js
Comment (ninaceban) [reply sandbox/src/App.js:66]: thank you

File: src/components/Personalization/decisionsExtractor.js
Comment (ninaceban) [reply src/components/Personalization/decisionsExtractor.js:33]: yes, I would love to improve that part

File: src/components/Personalization/createRedirectHandler.js
Comment (ninaceban) [reply src/components/Personalization/createRedirectHandler.js]: thank you

File: src/components/Personalization/decisionsExtractor.js
Comment (jonsnyder) [reply src/components/Personalization/decisionsExtractor.js:33]: https://jira.corp.adobe.com/browse/PDCL-5983

**Review by jonsnyder — APPROVED** [review]

**Pull request merged**
